File: kern/plfs.py
"""PLFS v0.1 flat catalog filesystem (S7b)."""

from __future__ import annotations

from dataclasses import dataclass

from kern.vfdd import VfddDriver
from plover_vm.memory.vfdd import SECTOR_SIZE


DIR_SECTOR = 1
DATA_BASE = 2

ENTRY_SIZE = 16
ENTRIES_PER_SECTOR = SECTOR_SIZE // ENTRY_SIZE


def _norm_name(name: str) -> bytes:
    # 8.3 upper, padded with spaces to 11 bytes.
    base = name.upper()
    if "." in base:
        a, b = base.split(".", 1)
    else:
        a, b = base, ""
    a = (a[:8]).ljust(8)
    b = (b[:3]).ljust(3)
    return (a + b).encode("ascii", errors="replace")


@dataclass(frozen=True)
class DirEntry:
    name11: bytes
    start_sector: int
    size_bytes: int
    attr: int = 0

    def pack(self) -> bytes:
        return (
            self.name11[:11]
            + bytes([self.start_sector & 0xFF, (self.start_sector >> 8) & 0xFF])
            + bytes([self.size_bytes & 0xFF, (self.size_bytes >> 8) & 0xFF])
            + bytes([self.attr & 0xFF])
        ).ljust(ENTRY_SIZE, b"\x00")

    @classmethod
    def unpack(cls, data: bytes) -> "DirEntry | None":
        if len(data) != ENTRY_SIZE:
            raise ValueError("bad entry size")
        name = data[:11]
        if name == b"\x00" * 11 or name == b" " * 11:
            return None
        start = data[11] | (data[12] << 8)
        size = data[13] | (data[14] << 8)
        attr = data[15]
        return cls(name11=name, start_sector=start, size_bytes=size, attr=attr)


class Plfs:
    def __init__(self, drv: VfddDriver) -> None:
        self.drv = drv

    def format(self) -> None:
        # sector0 reserved for future BPB; zero directory sector.
        self.drv.write_sector(DIR_SECTOR, b"\x00" * SECTOR_SIZE)

    def _load_dir(self) -> list[DirEntry | None]:
        sec = self.drv.read_sector(DIR_SECTOR)
        out: list[DirEntry | None] = []
        for i in range(ENTRIES_PER_SECTOR):
            chunk = sec[i * ENTRY_SIZE : (i + 1) * ENTRY_SIZE]
            out.append(DirEntry.unpack(chunk))
        return out

    def _store_dir(self, entries: list[DirEntry | None]) -> None:
        buf = bytearray(SECTOR_SIZE)
        for i, e in enumerate(entries[:ENTRIES_PER_SECTOR]):
            if e is None:
                continue
            buf[i * ENTRY_SIZE : (i + 1) * ENTRY_SIZE] = e.pack()
        self.drv.write_sector(DIR_SECTOR, bytes(buf))

    def list(self) -> list[DirEntry]:
        return [e for e in self._load_dir() if e is not None]

    def _find(self, name: str) -> tuple[int, DirEntry] | None:
        want = _norm_name(name)
        for i, e in enumerate(self._load_dir()):
            if e is None:
                continue
            if e.name11 == want:
                return i, e
        return None

    def create(self, name: str, data: bytes) -> None:
        if self._find(name) is not None:
            raise FileExistsError(name)
        entries = self._load_dir()
        try:
            slot = entries.index(None)
        except ValueError:
            raise RuntimeError("directory full")

        # Find first free data sector after existing files (contiguous policy).
        used_end = DATA_BASE
        for e in entries:
            if e is None:
                continue
            end = e.start_sector + ((e.size_bytes + SECTOR_SIZE - 1) // SECTOR_SIZE)
            used_end = max(used_end, end)
        start = used_end
        nsectors = (len(data) + SECTOR_SIZE - 1) // SECTOR_SIZE
        for s in range(nsectors):
            chunk = data[s * SECTOR_SIZE : (s + 1) * SECTOR_SIZE].ljust(SECTOR_SIZE, b"\x00")
            self.drv.write_sector(start + s, chunk)

        entries[slot] = DirEntry(name11=_norm_name(name), start_sector=start, size_bytes=len(data))
        self._store_dir(entries)

    def read(self, name: str) -> bytes:
        found = self._find(name)
        if found is None:
            raise FileNotFoundError(name)
        _, e = found
        nsectors = (e.size_bytes + SECTOR_SIZE - 1) // SECTOR_SIZE
        buf = bytearray()
        for s in range(nsectors):
            buf.extend(self.drv.read_sector(e.start_sector + s))
        return bytes(buf[: e.size_bytes])

    def delete(self, name: str) -> None:
        found = self._find(name)
        if found is None:
            raise FileNotFoundError(name)
        idx, _e = found
        entries = self._load_dir()
        entries[idx] = None
        self._store_dir(entries)

File: crates/plover_os/src/plfs.rs
use crate::vfdd::VfddDriver;
use plover_copro::SECTOR_SIZE;

pub const DIR_SECTOR: usize = 1;
pub const DATA_BASE: usize = 2;
pub const ENTRY_SIZE: usize = 16;
pub const ENTRIES_PER_SECTOR: usize = SECTOR_SIZE / ENTRY_SIZE;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct DirEntry {
    pub name11: [u8; 11],
    pub start_sector: u16,
    pub size_bytes: u16,
    pub attr: u8,
}

fn norm_name(name: &str) -> [u8; 11] {
    let upper = name.to_ascii_uppercase();
    let (a, b) = match upper.split_once('.') {
        Some((x, y)) => (x, y),
        None => (upper.as_str(), ""),
    };
    let mut out = [b' '; 11];
    for (i, c) in a.chars().take(8).enumerate() {
        out[i] = c as u8;
    }
    for (i, c) in b.chars().take(3).enumerate() {
        out[8 + i] = c as u8;
    }
    out
}

impl DirEntry {
    fn pack(&self) -> [u8; ENTRY_SIZE] {
        let mut buf = [0u8; ENTRY_SIZE];
        buf[..11].copy_from_slice(&self.name11);
        buf[11] = (self.start_sector & 0xFF) as u8;
        buf[12] = (self.start_sector >> 8) as u8;
        buf[13] = (self.size_bytes & 0xFF) as u8;
        buf[14] = (self.size_bytes >> 8) as u8;
        buf[15] = self.attr;
        buf
    }

    fn unpack(data: &[u8]) -> Option<Self> {
        if data.len() != ENTRY_SIZE {
            return None;
        }
        let mut name11 = [0u8; 11];
        name11.copy_from_slice(&data[..11]);
        if name11 == [0u8; 11] || name11 == [b' '; 11] {
            return None;
        }
        let start = u16::from(data[11]) | (u16::from(data[12]) << 8);
        let size = u16::from(data[13]) | (u16::from(data[14]) << 8);
        Some(DirEntry {
            name11,
            start_sector: start,
            size_bytes: size,
            attr: data[15],
        })
    }
}

pub struct Plfs {
    pub drv: VfddDriver,
}

#[derive(Debug)]
pub enum PlfsError {
    Exists,
    NotFound,
    DirFull,
    Vfdd(plover_copro::vfdd::VfddError),
}

impl From<plover_copro::vfdd::VfddError> for PlfsError {
    fn from(e: plover_copro::vfdd::VfddError) -> Self {
        PlfsError::Vfdd(e)
    }
}

impl Plfs {
    pub fn new(drv: VfddDriver) -> Self {
        Self { drv }
    }

    pub fn format(&mut self) -> Result<(), PlfsError> {
        self.drv
            .write_sector(DIR_SECTOR, &vec![0u8; SECTOR_SIZE])?;
        Ok(())
    }

    fn load_dir(&self) -> Result<Vec<Option<DirEntry>>, PlfsError> {
        let sec = self.drv.read_sector(DIR_SECTOR)?;
        let mut out = Vec::with_capacity(ENTRIES_PER_SECTOR);
        for i in 0..ENTRIES_PER_SECTOR {
            let chunk = &sec[i * ENTRY_SIZE..(i + 1) * ENTRY_SIZE];
            out.push(DirEntry::unpack(chunk));
        }
        Ok(out)
    }

    fn store_dir(&self, entries: &[Option<DirEntry>]) -> Result<(), PlfsError> {
        let mut buf = vec![0u8; SECTOR_SIZE];
        for (i, e) in entries.iter().take(ENTRIES_PER_SECTOR).enumerate() {
            if let Some(ent) = e {
                buf[i * ENTRY_SIZE..(i + 1) * ENTRY_SIZE].copy_from_slice(&ent.pack());
            }
        }
        self.drv.write_sector(DIR_SECTOR, &buf)?;
        Ok(())
    }

    pub fn list(&self) -> Result<Vec<DirEntry>, PlfsError> {
        Ok(self
            .load_dir()?
            .into_iter()
            .flatten()
            .collect())
    }

    fn find(&self, name: &str) -> Result<Option<(usize, DirEntry)>, PlfsError> {
        let want = norm_name(name);
        for (i, e) in self.load_dir()?.into_iter().enumerate() {
            if let Some(ent) = e {
                if ent.name11 == want {
                    return Ok(Some((i, ent)));
                }
            }
        }
        Ok(None)
    }

    pub fn create(&mut self, name: &str, data: &[u8]) -> Result<(), PlfsError> {
        if self.find(name)?.is_some() {
            return Err(PlfsError::Exists);
        }
        let mut entries = self.load_dir()?;
        let slot = entries
            .iter()
            .position(|e| e.is_none())
            .ok_or(PlfsError::DirFull)?;

        let mut used_end = DATA_BASE;
        for e in entries.iter().flatten() {
            let end = e.start_sector as usize
                + ((e.size_bytes as usize + SECTOR_SIZE - 1) / SECTOR_SIZE);
            used_end = used_end.max(end);
        }
        let start = used_end;
        let nsectors = (data.len() + SECTOR_SIZE - 1) / SECTOR_SIZE;
        for s in 0..nsectors {
            let off = s * SECTOR_SIZE;
            let mut chunk = vec![0u8; SECTOR_SIZE];
            let end = (off + SECTOR_SIZE).min(data.len());
            chunk[..end - off].copy_from_slice(&data[off..end]);
            self.drv.write_sector(start + s, &chunk)?;
        }

        entries[slot] = Some(DirEntry {
            name11: norm_name(name),
            start_sector: start as u16,
            size_bytes: data.len() as u16,
            attr: 0,
        });
        self.store_dir(&entries)
    }

    pub fn read(&self, name: &str) -> Result<Vec<u8>, PlfsError> {
        let (_, e) = self.find(name)?.ok_or(PlfsError::NotFound)?;
        let nsectors = (e.size_bytes as usize + SECTOR_SIZE - 1) / SECTOR_SIZE;
        let mut buf = Vec::new();
        for s in 0..nsectors {
            buf.extend_from_slice(&self.drv.read_sector(e.start_sector as usize + s)?);
        }
        buf.truncate(e.size_bytes as usize);
        Ok(buf)
    }

    pub fn delete(&mut self, name: &str) -> Result<(), PlfsError> {
        let (idx, _) = self.find(name)?.ok_or(PlfsError::NotFound)?;
        let mut entries = self.load_dir()?;
        entries[idx] = None;
        self.store_dir(&entries)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use plover_copro::vfdd::{VfdConfig, VirtualFdd};
    use std::time::{SystemTime, UNIX_EPOCH};

    fn tmp_disk() -> (std::path::PathBuf, Plfs) {
        let t = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos();
        let path = std::env::temp_dir().join(format!("plover_plfs_{t}.img"));
        let dev = VirtualFdd::new(VfdConfig {
            path: path.clone(),
            sector_count: 64,
        })
        .unwrap();
        let mut fs = Plfs::new(VfddDriver::new(dev));
        fs.format().unwrap();
        (path, fs)
    }

    #[test]
    fn create_read_delete() {
        let (path, mut fs) = tmp_disk();
        fs.create("HELLO.TXT", b"hello").unwrap();
        assert_eq!(fs.read("HELLO.TXT").unwrap(), b"hello");
        assert_eq!(fs.list().unwrap()[0].name11, *b"HELLO   TXT");
        fs.delete("HELLO.TXT").unwrap();
        assert!(fs.list().unwrap().is_empty());
        let _ = std::fs::remove_file(path);
    }
}
